backend/store/user_group.go

```diff
@@ -0,0 +1,268 @@
+package store
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"google.golang.org/protobuf/encoding/protojson"
+
+	"github.com/pkg/errors"
+
+	"github.com/bytebase/bytebase/backend/common"
+	storepb "github.com/bytebase/bytebase/proto/generated-go/store"
+)
+
+// FindUserGroupMessage is the message for finding groups.
+type FindUserGroupMessage struct {
+	Email *string
+}
+
+// UpdateUserGroupMessage is the message to update a group.
+type UpdateUserGroupMessage struct {
+	Title       *string
+	Description *string
+	Payload     *storepb.UserGroupPayload
+}
+
+// UserGroupMessage is the message for a group.
+type UserGroupMessage struct {
+	Email       string
+	Title       string
+	Description string
+	CreatorUID  int
+	Payload     *storepb.UserGroupPayload
+}
+
+// GetUserGroup gets a group.
+func (s *Store) GetUserGroup(ctx context.Context, email string) (*UserGroupMessage, error) {
+	if v, ok := s.userGroupCache.Get(email); ok {
+		return v, nil
+	}
+
+	tx, err := s.db.BeginTx(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+	defer tx.Rollback()
+
+	groups, err := s.listUserGroupImpl(ctx, tx, &FindUserGroupMessage{
+		Email: &email,
+	})
+	if err != nil {
+		return nil, err
+	}
+	if err := tx.Commit(); err != nil {
+		return nil, err
+	}
+
+	if len(groups) == 0 {
+		return nil, nil
+	} else if len(groups) > 1 {
+		return nil, &common.Error{Code: common.Conflict, Err: errors.Errorf("found %d groups with email %+v, expect 1", len(groups), email)}
+	}
+	return groups[0], nil
+}
+
+// ListUserGroups list all groups.
+func (s *Store) ListUserGroups(ctx context.Context, find *FindUserGroupMessage) ([]*UserGroupMessage, error) {
+	tx, err := s.db.BeginTx(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+	defer tx.Rollback()
+
+	groups, err := s.listUserGroupImpl(ctx, tx, find)
+	if err != nil {
+		return nil, err
+	}
+	if err := tx.Commit(); err != nil {
+		return nil, err
+	}
+
+	for _, group := range groups {
+		s.userGroupCache.Add(group.Email, group)
+	}
+	return groups, nil
+}
+
+func (*Store) listUserGroupImpl(ctx context.Context, tx *Tx, find *FindUserGroupMessage) ([]*UserGroupMessage, error) {
+	where, args := []string{"TRUE"}, []any{}
+	if v := find.Email; v != nil {
+		where, args = append(where, fmt.Sprintf("email = $%d", len(args)+1)), append(args, *v)
+	}
+
+	var groups []*UserGroupMessage
+	rows, err := tx.QueryContext(ctx, fmt.Sprintf(`
+	SELECT
+		email,
+		creator_id,
+		name,
+		description,
+		payload
+	FROM user_group
+	WHERE %s
+	`, strings.Join(where, " AND ")), args...)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var group UserGroupMessage
+		var payload []byte
+		if err := rows.Scan(
+			&group.Email,
+			&group.CreatorUID,
+			&group.Title,
+			&group.Description,
+			&payload,
+		); err != nil {
+			return nil, err
+		}
+
+		groupPayload := storepb.UserGroupPayload{}
+		decoder := protojson.UnmarshalOptions{DiscardUnknown: true}
```

Comment:
We don't need `DiscardUnknown: true` since it's a fresh start.

Reply:
LGTM, updated